src/Core/PolynomialFunction.java

```diff
@@ -129,31 +132,98 @@ public void derive() {
         functionString =  sb.toString();
     }
 
-    public void calcRoots(double start, double end, double step) {
-        roots = new ArrayList<>();
-        Vector2D prevPoint = evaluate(start);
+    public void calcRootsAndExtremes(double start, double end, double step) {
+        // If function is a constant skip the calculation
+        if(coefficients.length == 1) {
+            roots = new HashSet<>();
+            extremePoints = new HashSet<>();
+            roots.add(Double.NaN);
+            extremePoints.add(Double.NaN);
+            return;
+        }
+        Callable<Set<Double>> task1 = () -> calcRoots(start, end, step);
+        Callable<Set<Double>> task2 = () -> calcExtremes(start, end, step);
+
+        try {
+            List<Future<Set<Double>>> results = executorService.invokeAll(Arrays.asList(task1, task2));
+            roots = results.get(0).get();
+            extremePoints = results.get(1).get();
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static final double EPSILON = 1E-15;
```

Comment:
please move this to the other member variable declarations or declare it locally where used inside the function

Reply:
I thougth i could hide it because you hate getting more and more variable declarations and declaring it locally wouldnt really be the best choice :D

Reply:
Done